pkg/stanza/fileconsumer/reader_factory.go

```diff
@@ -122,7 +128,8 @@ func (b *readerBuilder) build() (r *Reader, err error) {
 		r.Fingerprint = fp
 	}
 
-	if !b.fromBeginning {
+	// unsafeReader has the file set to nil, so don't try emending its offset.
+	if !b.fromBeginning && !b.isUnsafe {
```

Comment:
Technically, we don't need the additional variable and associated code, and could just move this block into the `if b.file != nil` condition above. I think the comment you've added would be sufficient to explain why it is located there.